### DISCUSSION TITLE
GitHub Action using ImageMagick: Create a PNG and commit it to the repo

### DISCUSSION THREAD

**Discussion by KasparJohannesSchneider** [community/community#26600] [Actions]
<p><em>I asked this question earlier today on <a href="https://stackoverflow.com/questions/67060829/github-action-using-imagemagick-create-a-png-and-commit-it-to-the-repo" rel="noopener nofollow ugc">Stack Overflow</a>, however, I think it fits better here…</em></p>
<p><strong>Background</strong><br>
I’m maintaining a forked <a href="https://github.com/KasparJohannesSchneider/TwentySecondsCurriculumVitae-LaTex" rel="noopener nofollow ugc">repository</a> that contains a LaTeX project. The <code>README.md</code> contains a <code>.png</code> preview of the <code>.pdf</code>, compiled from the example <code>.tex</code> file included in the repo. I often forget to update the <code>.png</code> version using the following ImageMagick command:</p>
<pre><code class="lang-auto">    magick convert -density 1200 -background white -alpha off Twenty-Seconds-Icons_cv.pdf -quality 90 Twenty-Seconds-Icons_cv.png
</code></pre>
<p>Thus I would like to automate this process using GitHub Actions.</p>
<p><strong>Workflow</strong><br>
I think the <code>main.yml</code> file should look something like this, however, I don’t fully understand what I am doing.</p>
<pre><code class="lang-auto">name: PDF to PNG
on:
  push:
    branches:
      - kaspar
  pull_request:
    branches:
      - kaspar

jobs:
  build:
    runs-on: ubuntu-latest
    steps:
      - name: Install ImageMagick
        run: sudo apt install imagemagick # Seems to work and already be included with ubuntu-latest...
      - name: Convert PDF to PNG
        run: # How?
      - name: Commit the updated PNG 
        run: # How?
</code></pre>
<p><strong>Output:</strong></p>
<pre><code class="lang-auto">Run sudo apt install imagemagick
  sudo apt install imagemagick
  shell: /usr/bin/bash -e {0}

WARNING: apt does not have a stable CLI interface. Use with caution in scripts.

Reading package lists...
Building dependency tree...
Reading state information...
imagemagick is already the newest version (8:6.9.10.23+dfsg-2.1ubuntu11.2).
0 upgraded, 0 newly installed, 0 to remove and 0 not upgraded.
</code></pre>
<p><strong>Questions</strong><br>
It seems like this should be a pretty basic task, therefore, I would prefer to do everything from the <code>main.yml</code> file without creating a separate repository for this action, if possible.</p>
<ol>
<li>It seems like I can just run bash shell commands with <code>run: </code>; is this true?</li>
<li>How/where can I access the <code>.pdf</code> file in the repo from inside the GitHub Actions Ubuntu machine?</li>
<li>Where should I save the <code>.png</code> file, is there a <code>~</code> directory?</li>
<li>How can I commit the <code>.png</code> file generated by ImageMagick to the repo?</li>
<li>Would <code>git commit &lt;file&gt; -m "updated png version of the CV"</code> work?</li>
</ol>

**Comment by KasparJohannesSchneider** [comment, marked as answer]
<p>It seems like I figured out the/a solution myself.<br>
<strong>I would be very thankful for feedback (code review)  since my experience with GitHub Actions is very limited…</strong></p>
<pre><code class="lang-yaml">name: PDF to PNG
on:
  push:
    branches:
      - kaspar
  pull_request:
    branches:
      - kaspar

jobs:
  build:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v2
      - name: Install ghostscript
        run: sudo apt install ghostscript
      - name: Change ImageMagick security policy
        run: |
          DQT='"' 
          SRC="rights=${DQT}none${DQT} pattern=${DQT}PDF${DQT}"
          RPL="rights=${DQT}read\|write${DQT} pattern=${DQT}PDF${DQT}"
          sudo sed -i "s/$SRC/$RPL/" /etc/ImageMagick-6/policy.xml
      - name: Convert PDF to PNG
        run: convert -density 900 -background white -alpha off Twenty-Seconds-Icons_cv.pdf -quality 90 Twenty-Seconds-Icons_cv.png
      - name: Commit PNG
        id: commit
        run: |
          git config --local user.email "action[bot]@github.com"
          git config --local user.name "github-actions[bot]"
          git add Twenty-Seconds-Icons_cv.png
          if [-z "$(git status --porcelain)"]; then
             echo "::set-output name=push::false"
          else
             git commit -m "[bot] updated Twenty-Seconds-Icons_cv.png"
             echo "::set-output name=push::true"
          fi
        shell: bash
      - name: Push Commit
        if: steps.commit.outputs.push == 'true'
        uses: ad-m/github-push-action@master
        with:
           github_token: ${{ secrets.SECRET_TOKEN }}
</code></pre>
<p><strong>What I was doing wrong:</strong></p>
<ul>
<li>I did not checkout the repository</li>
<li>ghostscript was missing :ghost:
</li>
<li>The security policy for ImageMagic needs to be edited for it to process PDFs :angry:
</li>
<li>On Ubuntu the ImageMagick command doesn’t work for 1200 dpi (900 is ok in my case), on Windows, this works just fine  :man_shrugging:t2:
</li>
<li>I found out <a href="https://github.community/t/how-to-make-a-workflow-file-to-save-the-output-as-a-file-to-my-github-repo/18352/2">how to commit and push</a>
</li>
</ul>

**Comment by junguler** [comment]
i was going insane trying to figure out how to use imagemagick in github actions, literally hours wasted and 100 of failed commits/runs in private repos, this really helped me so thank you
